### ISSUE TITLE
Module Installation broken

### ISSUE THREAD

**Issue by MaxE17677** [Smile-SA/elasticsuite#175]
Magento 2 Module Install

I follow the description at  the [wiki](https://github.com/Smile-SA/elasticsuite/wiki/ModuleInstall) how to install the module. 

But `bin/magento setup:upgrade` fails on two points because my elasticsearch is not availible at localhost:9200.

```php
// Smile\ElasticsuiteCatalog\Setup\InstallData Line 95
$this->getIndexer('elasticsuite_categories_fulltext')->reindexAll(); 
```
```php
// Smile\ElasticsuiteThesaurus\Setup\InstallData Line 57
$this->getIndexer('elasticsuite_thesaurus')->reindexAll();
```

When i remove this two lines, every thing went fine. 

Then i can log in. I can change the host and port like documented and start indexing.

If there is a way to change host and port befor run `bin/magento setup:upgrade` please document.

**Comment by romainruaud** [comment]
Hello @MaxE17677 and thank you for pointing us this issue.

We have plenty of projects running with the module on UAT and production environments which use a dedicated ElasticSearch server (not on localhost:9200). I will ask our teams to know how they are doing and I'll come back to tell you.

For now, I think they may edit manually the config.xml file of the module to specify the proper ElasticSearch address they are using, or they may have another module in their project dedicated code pool which is overriding the Elasticsuite config.xml with the proper server location.

I keep you in touch,

Regards,

**Comment by afoucret** [comment]
Modifying the config.xml file shipped with ElasticSuite is not acceptable regarding best practices because : 

* you can not update the module when you get this done

* it is not into your VCS since vendor are not commited

The best practice is to override this default configuration into your own module. An example would be to have a `MyProject_ElasticSuite` module in your own project (into `app/code`) with the following content into the `etc/config.xml` file :

```xml
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Store:etc/config.xsd">
    <default>
        <smile_elasticsuite_core_base_settings>
            <es_client>
                <enable_debug_mode>0</enable_debug_mode>
                <servers>localhost:9200</servers>
                <connection_timeout>1</connection_timeout>
                <timeout>30</timeout>
            </es_client>
        </smile_elasticsuite_core_base_settings>
    </default>
</config>
```

Then you will be able to put all your ElasticSuite override there.

@MaxE17677 : Does this solution is suitable for you ?

@romainruaud : can you add this as a note for installing on a separate serverinto the install wiki ?

**Comment by MaxE17677** [comment]
This helps a lot. Thank you @afoucret and @romainruaud.

**Comment by romainruaud** [comment]
Added to the Wiki : https://github.com/Smile-SA/elasticsuite/wiki/ModuleInstall#optional--installing-the-module-when-using-a-custom-elasticsearch-server-location

Regards.

**Comment by JamesonNetworks** [comment]
This can still happen. Essentially, if you have ElasticSearch configured, but you've selected a different search provider like MySQL in M2, and then you go to install this module, it picks up the configured elastic search url and attempts to use that to connect. When it can't connect, the module fails to install. I believe we should be checking the current search provider, and if it ISN'T elastic search, avoid the reindex altogether. 

Where I think you will start seeing this error a lot is with customers who are using the Magento 2 Cloud offering. I ran into it because I pulled a cloud database which had elastic search configured and it bombed out my category pages. I changed the config to MySQL search and everything worked fine. Then, several weeks later, I attempted to install this module which failed at the reindexing steps on install. Because another module after it had a db revision script to run, I was locked out of the admin with the error messages about data version not matching. I installed elasticsearch as an attempt to get around the errors, however, the cloud offering uses a special DNS name called elastic.internal. Since it was not localhost, the module still failed to install.

TL;DR:
The install script should be checking the current search provider, and if it ISN'T elastic search, avoid the reindex altogether as to not lock people out of the admin.
